### PR TITLE
A big update pass on the contributing guide.

### LINKED ISSUE #159: Do not run full CI by default, require full CI before merging.
This was discussed during a Coq Call in late 2020: https://github.com/coq/coq/wiki/Coq-Call-2020-11-25

The idea was to distinguish between jobs that would be always run (lightweight mode) and jobs that would only be run on request (full mode). To ensure that we don't break the CI in `master` we should still require the full mode to be run on all PRs before merging. This could be ensured with a specific status check.

This has become less critical now and it's not clear that we still want this after:
- we got a Gold OSS plan on GitLab, lifting the 500 job limit;
- the auto-minimizer was added, which makes it much more valuable to get all the jobs to be run early (that being said, if we can get the test-suite to become complete enough to cover most failure cases, then this should become useful once more).

FWIW if / when we implement this feature, we could rely on the push option support of GitLab to specify if we are in lightweight mode or in full mode every time a PR is updated: https://docs.gitlab.com/ee/user/project/push_options.html#push-options-for-gitlab-cicd

### PR DESCRIPTION
**Pull request by Zimmi48**
See the list of commits for details about the changes.

Closes https://github.com/coq/bot/issues/159.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -326,10 +326,11 @@ the `find-bug.py` script in that repo).
...
 
-Coq gets a new major release about every six months.  Before a new
-major version is released, there is a beta-testing period, which
-usually lasts one month (see the [release plan][release-plan]).  You
-can help make the upcoming release better, by testing the beta
+Coq gets a new major release about every six months, which is then
+distributed through the [Coq Platform][Coq-Platform].
+New major versions of Coq are first made available for beta-testing,
+before being declared stable and the new default version of the Coq
+Platform. You can help make by testing the beta
```
Comment (Alizter) [inline CONTRIBUTING.md:335]: No mention of opam?

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -539,9 +539,22 @@ commit merging your branch with the base branch, so if there is a
...
+Otherwise, a suite of tests will be run on GitLab, plus some
 additional tests on GitHub Actions for Windows and macOS compatibility.
 
+The complete suite of tests is no longer run by default to save
+resources. But it is still required before merging a PR, so this is why
+the bot will put a `needs: full CI` label if it has only run the
+lightweight tests. If you are a member of **@coq/contributors**, you can
+request a full run of the CI by putting the `request: full CI` label
+before pushing to your PR branch, or by commenting
+`@coqbot: run full CI` after having pushed. (In case you need to re-run
```
Comment (Alizter) [inline CONTRIBUTING.md:552]: Let's keep the colons consistent. I think not having them is better.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -539,9 +539,22 @@ commit merging your branch with the base branch, so if there is a
...
+the tests, e.g., because the results are outdated, you can also request
+the bot to do so by commenting `@coqbot: run full CI` or
+`@coqbot: run light CI`. If you comment `@coqbot run CI`, the bot will
```
Comment (Alizter) [inline CONTRIBUTING.md:555]: Here the colon differs.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -677,18 +679,23 @@ Here are a few labels that reviewers may add to your PR to track its
...
   changes that should be documented in the [user
-  changelog][user-changelog].
+  changelog][user-changelog]. Similarly to the previous label, this
+  may be used to reflect that the corresponding checkbox is not yet
+  checked in the PR template.
+- [needs: test-suite update][needs-test-suite] indicates that tests
+  should be added to the test-suite to ensure that the changes are
+  properly tested. Similarly to the previous two labels, this may be
+  used to reflect that the corresponding checkbox is not yet checked
+  in the PR template.
```
Comment (Alizter) [inline CONTRIBUTING.md:698]: Not just adding tests but fixing them too.

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -412,10 +412,6 @@ Being in this team will grant you the following access:
   community moderators.  You should hide them instead.  An audit log
   is available to track deleted items if needed (but does not allow
   recovering them).
-- **Pushing a branch or a tag to the main repository:** please push
-  changes to your own fork rather than the main repository. (Branches
-  pushed to the main repository will be removed promptly and without
-  notice.)
```
Comment (SkySkimmer) [inline CONTRIBUTING.md:397]: Shouldn't we say "pushing branches to the main repo is disabled" rather than say nothing?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -840,6 +840,14 @@ reviewer for each component that is significantly affected), that
 change requests have been implemented, that CI is passing, and
 eventually will be the one who merges the PR.
 
+The PR assignee may use their own judgement to decide to merge a PR that
+has not received reviews from all maintainers of affected components,
+depending on how large or controversial the changes to these components
+are. It is also admitted to have an assignee who is not a maintainer of
```
Comment (SkySkimmer) [inline CONTRIBUTING.md:852]: "admitted" doesn't sound right imo

File: CONTRIBUTING.md (hunk 2 above)
Comment (Alizter) [reply CONTRIBUTING.md:552]: This is about the `@coqbot` commands.

File: CONTRIBUTING.md (hunk 6 above)
Comment (Alizter) [reply CONTRIBUTING.md:852]: I think `admissible` is probably what was intended.

File: CONTRIBUTING.md (hunk 1 above)
Comment (Zimmi48) [reply CONTRIBUTING.md:335]: What kind of mention would you expect / like to see here?

File: CONTRIBUTING.md (hunk 2 above)
Comment (Zimmi48) [reply CONTRIBUTING.md:552]: OK. I'll remove them.

File: CONTRIBUTING.md (hunk 1 above)
Comment (Alizter) [reply CONTRIBUTING.md:335]: Well reading this the first time, it seems that it is only distributed through the Coq platform, which of course isn't true. However I suppose it isn't that important to mention here really.

**Comment by Zimmi48** [comment]
AFAICT this is ready to merge.

**Comment by Zimmi48** [comment]
I believe that's spurious, because CONTRIBUTING is not used in the build process, but re-running to be safe.

**Review by Alizter — APPROVED** [review]

**Comment by Alizter** [comment]
@coqbot merge now

**Comment by Alizter** [comment]
@coqbot merge now

**Pull request merged**